errors/error.go

```diff
@@ -0,0 +1,268 @@
+				}
+			}
+
+		case Stack:
+			e.Stack = arg
+		case string:
+			e.Description = arg
+		case error:
+			e.Err = arg
+		default:
+			panic(fmt.Errorf("called with unknown type %T", arg))
+		}
+	}
+	if e.isEmpty() {
+		panic(fmt.Errorf("empty error"))
+	}
+	prev, ok := e.Err.(*Error)
+	if !ok {
+		return e
+	}
+	if prev.Kind == e.Kind {
+		prev.Kind = ""
+	} else if e.Kind == "" {
+		e.Kind = prev.Kind
+		prev.Kind = ""
+	}
+	if e.FileRange.Empty() {
+		e.FileRange = prev.FileRange
+	}
+	if prev.FileRange == e.FileRange {
+		prev.FileRange = hcl.Range{}
+	}
+	if prev.Stack == e.Stack {
+		prev.Stack = ""
+	}
+
+	if prev.Description == e.Description {
+		prev.Description = ""
+	}
+
+	if prev.isEmpty() {
+		e.Err = prev.Err
+	}
+
+	// If this error has Kind unset or Other, pull up the inner one.
+	if e.Kind == "" {
+		e.Kind = prev.Kind
+		prev.Kind = ""
+	}
+	return e
+}
+
+// isEmpty tells if all fields of this error are empty.
+// Note that e.Err is the underlying error hence not checked.
+func (e *Error) isEmpty() bool {
+	return e.FileRange.Empty() && e.Kind == "" && e.Description == "" && e.Stack == ""
+}
+
+// Error returns the error message.
+func (e *Error) Error() string {
+	if e.isEmpty() {
+		if e.Err != nil {
+			return e.Err.Error()
+		}
+		return ""
+	}
+
+	var errParts []string
+	for _, arg := range []interface{}{
+		e.FileRange,
+		e.Kind,
+		e.Description,
+		e.Stack,
+		e.Err,
+	} {
+		switch v := arg.(type) {
+		case hcl.Range:
+			if !v.Empty() {
+				errParts = append(errParts, v.String())
+			}
+		case Kind:
+			if v != "" {
+				errParts = append(errParts, string(v))
+			}
+		case string:
+			if v != "" {
+				errParts = append(errParts, v)
+			}
+		case Stack:
+			if v != "" {
+				errParts = append(errParts, string(v))
+			}
+		case error:
+			if v != nil {
+				errParts = append(errParts, v.Error())
+			}
+		case nil:
+			// ignore nil values
+		default:
+			panic(fmt.Sprintf("unexpected case: %+v", arg))
+		}
+	}
+
+	return strings.Join(errParts, separator)
+}
+
+// IsKind tells if err is of kind k.
+// It returns false if err is nil or not an *errors.Error.
+// It also recursively checks if any underlying error is of kind k.
+func IsKind(err error, k Kind) bool {
+	if err == nil {
+		return false
+	}
+	e, ok := err.(*Error)
+	if !ok {
+		return false
+	}
+	if e.Kind != "" {
+		return e.Kind == k
+	}
+	return IsKind(e.Err, k)
+}
+
+// Is tells if err (or any of its underlying errors) matches target.
+// It works with any error but if comparing *Error type it uses the Kind field,
+// otherwise it fallback to standard errors.Is().
+func Is(err, target error) bool {
+	e, ok := err.(*Error)
+	if !ok {
+		return errors.Is(err, target)
+	}
+	t, ok := target.(*Error)
+	if !ok {
+		// target is not *Error so it's already different.
+		// Let's check the underlying error.
+		if e.Err != nil {
+			return Is(e.Err, target)
+		}
+		return false
+	}
+
+	// both are *Error
+	if IsKind(e, t.Kind) {
+		return true
+	}
```

Comment:
hmm so errors of the same kind but for different stacks would be true when checking errors.Is() ? :thinking: 
we could ignore fields that are not present on the target error, so it would be a partial match, but if both errors have kind + stack and stack differs it seems odd to detect as true, or maybe Im missing something here.

Reply:
I thought about this but the same thing could be said about the other metadata/info (filerange, description, etc). Currently (I mean what we have in `main` now) the stack info goes in the error message so we also don't check this. 

We can add this improvement in a separate PR together with the tests changes to make use of it, but I think everything available in the `target` should be used for match. Just don't think make sense adding this now because I'll have to work in the tests to use them, otherwise it's just dead code.